waitinterface.1
.TH waitinterface 1
.SH NAME
waitinterface \- wait for a network interface to show up
.SH SYNOPSIS
.B waitinterface ifname command ...

.B waitinterfaceup ifname command ...
.SH DESCRIPTION
.B waitinterface
will wait for /sys/class/net/ifname to show up (it will try for a few
seconds, then abort and fail if nothing shows up).

If it is called under the name waitinterfaceup (make a symbolic or hard
link at installation!), it will also wait for the interface to go up. If
that does not happen within a few seconds, it will abort and fail.

If execution gets this far, the rest of the command line is run (using
execve, so you have to pass in the full path to the command).

This program is meant for use when one command loads a network driver
into the kernel, and the next program wants to initialize it. There is a
race condition between the kernel module creating the network interface
and the initialization functions attempting to initialize that
interface. In this case you can use waitinterface to delay the
initialization until the interface is actually there.

.SH AUTHOR
waitinterface was written by Felix von Leitner and can be downloaded from
.I http://www.fefe.de/minit/



CHANGES
0.11:
  when serdo is run without arguments, and a file called "script"
    exists in the current working directory, executes that (to save a
    few inodes for the "params" files
  fix (apparently unexploitable) stack clobber in serdo
    (found by Nikola Vladov, many thanks)
  add ulimit to serdo
  a service can now contain a "root" symlink to where it should be
    started
  only install SIGWINCH and SIGINT if running with PID 1
  you can now do "make MINITROOT=/var/minit"
  add ftrigger (run a command if a file changes)
  add waitinterface and waitport (to resolve races)
  add powersave (demo tool to set up power saving, needs customization, not actually installed)
  add governor (demo tool to set CPU frequency governor based on AC state,
    not actually needed anymore with current hardware, not actually installed)

0.10:
  add sample script for /etc/minit/ctrlaltdel/run as
    control/ctrlaltdel-run (Florian Westphal)
  add killall5, needs /proc on Linux though (Florian Westphal)
  add shutdown (Bernd Wachter)
  add -Wall and -W and kill resulting compiler warnings
  several fixes (Bernd Wachter)
  add update support - tell minit to execve something else, a newer
    version of itself, preferably (Florian Westphal)
  also add minit-update program to download state from old minit and
    upload state to new minit after update (Florian Westphal)
  #include minit.h and use MINITROOT consistently (Lordy)
  Don't assume MINITROOT is 10 bytes long (Erich Schubert)
  add man pages!! (Erich Schubert)
  add a way to ask minit for dependencies
  msvc -H will now dump the ten least recently spawned processes.
    This is useful to see which process is looping.
  fix bug in msvc killing process group if readpid failed
    (Erich Schubert)
  You can now blacklist services temporarily by passing -servicename on
    the kernel command line.  The first time they are to be started will
    then be skipped.  Use this, for example, to not start the dhcp
    client when your notebook is not plugged in a network.
  add serdo (execute batch, no shell features whatsoever)
  make "msvc servicename" output easier to parse if not outputting to a tty
  require libowfat instead of shipping half of it ourselves
  replace malloc with alloca; minit is now 6644 bytes on i386.
  add usage for serdo (Nikola Vladov)
  fix msvc -d followed by msvc -h (would send SIGHUP to pid 1, i.e. kill
    minit; found by Paul Fox)
  added small special-purpose malloc using a static buffer before
    calling the real malloc (Nikola Vladov)

0.9.1:
  fix embarassing typo in msvc (Gelu G. Lupas)

0.9:
  For stopped services, minit did not record the time when it finished,
    but when it was started.
  add hard-reboot from Tommi Virtanen.  He also asked me to include an
    explicit license statement and sent a patch to make debian packaging
    easier (deb files can't include FIFOs).
  Olaf: add write_proc.
  Olaf: lines in depends can now be commented out with a #
  Olaf: add a method to clear dead services (mark as terminated)
  minit now checks whether a PID actually exists before accepting msvc -P
    this removes a race condition when the forked service terminates
    before pidfilehack notifies minit of the PID
  msvc now accepts more than one service after -o, -d, ...
  msvc now accepts /etc/minit/sshd instead of sshd as service name
  better error handling in msvc (more and better error messages)
  minit now has a heuristic to detect time zone changes and still
    have msvc report the correct elapsed time since process start/end
    (don't look at the source code, it's ultra-kludgey)
  add msvc -g (get PID for shell scripts)

0.8:
  call waitpid repeatedly until it returns "no children".
  This reaps zombies faster.

0.7:
  found and fixed the bug that made the logging support not work when
    minit was running as PID 1.
  saw that the shutdown from sysvinit actually does what it's supposed
    to do, even with minit, if one uses the -n mode.  It should be quite
    easy to write a nice and clean shutdown for minit now.
  if a service depends on a service that already finished, don't start
    it again.

0.6.1:
  fixed msvc and minit handling of msvc -P.
  pidfilehack has now actually been tested and works with ssh.

0.6:
  add tty and session leadership magic and I actually got getty to work
    as init on my laptop, i.e. in production use!
  add msvc man page.

0.5.1:
  cut&paste error, set wrong variable to 0 for keyboard request.
    Thanks, Schulti.

0.5:
  add "log" subdirectory (see README).
  get rid of stdio and *printf and switch to routines from libowfat.

0.4:
  add pidfilehack.  It will take the first argument as filename for the
  PID file (i.e. /var/run/sshd.pid), the second argument as service name
  for minit (i.e. ssh) and the rest of the command line is run without
  path searching in a child process.  The parent process will then
  try to open the pidfile several times (for up to 30 seconds), read the
  PID off it and run msvc to tell minit the PID of the service.  minit
  can then track sshd and restart it if necessary.  This gross hack does
  not work for daemons that do not write a PID file, but there are not
  many of those, fortunately.


waitinterface.c
#include <unistd.h>
#include <errmsg.h>
#include <time.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <string.h>
#include <fmt.h>
#include <netinet/in.h>
#include <net/if.h>
#include <sys/ioctl.h>

int main(int argc,char* argv[],char* envp[]) {
  unsigned int i;
  struct timespec req,rem;
  struct stat ss;
  char* sys;
  int found;
  errmsg_iam("waitinterface");
  if (argc<2)
    die(0,"usage: waitinterface wlan0 /usr/sbin/dhcpcd wlan0\n\twaits for wlan0 to show up and then executes the rest of the command line");
  sys=fmt_strm_alloca("/sys/class/net/",argv[1]);
  req.tv_sec=0; req.tv_nsec=100000000;
  for (i=0; i<100; ++i) {
    if ((found=lstat(sys,&ss))==0) break;
    nanosleep(&req,&rem);
  }
  if (found==-1)
    die(1,"interface not showing up");
  if (strstr(argv[0],"up")) {
    /* they want the interface to be up, too */
    int s=socket(PF_INET6,SOCK_DGRAM,IPPROTO_IP);
    struct ifreq ifr;
    if (s==-1) s=socket(PF_INET,SOCK_DGRAM,IPPROTO_IP);
    if (s==-1) diesys(1,"socket");
    strncpy(ifr.ifr_name,argv[1],sizeof(ifr.ifr_name));
    for (i=0; i<100; ++i) {
      if (ioctl(s,SIOCGIFFLAGS,&ifr)==0 && (ifr.ifr_flags&IFF_UP)) break;
      nanosleep(&req,&rem);
    }
    close(s);
    if (!(ifr.ifr_flags&IFF_UP))
      die(1,"interface not up");
  }
  if (argv[2]) {
    execve(argv[2],argv+2,envp);
    diesys(1,"execve");
  }
  return 0;
}
